src/main/java/com/google/sps/servlets/GroupsServlet.java

```diff
@@ -21,21 +26,45 @@
      */
     @Override
     public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
-        // Hard coded data for now but will use data from datastore
-        // as groups are added to datastore
+        // Hard coded groups for now, but hopefully will use user-created groups from datastore
         // Photo by Hybrid on Unsplash 
         // (https://unsplash.com/@artbyhybrid?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText)
-        ImmutableMap<String, String> groupMap = ImmutableMap.of("groupName", "Black Lives Matter",
-            "groupImage", "https://images.unsplash.com/photo-1591622414912-34f2a8f8172e?ixlib=" + 
-            "rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80",
-            "groupDescription", "Group description would go here...");
-        ImmutableList<ImmutableMap<String, String>> groups = ImmutableList.of(groupMap);
+        Entity blmEntity = new Entity(GROUP_KIND, "123");
+        blmEntity.setProperty(GROUP_NAME_PROPERTY, "Black Lives Matter");
+        blmEntity.setProperty(GROUP_IMAGE_PROPERTY, "https://images.unsplash.com/photo-1591622414912-34f2a8f8172" + 
+            "e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1950&q=80");
+        blmEntity.setProperty(GROUP_DESCRIPTION_PROPERTY, "Group description would go here...");
+        datastore.put(blmEntity);
+        
+        // Photo by Conscious Design on Unsplash
+        // (https://unsplash.com/@conscious_design?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText)
+        Entity sierraEntity = new Entity(GROUP_KIND, "456");
+        sierraEntity.setProperty(GROUP_NAME_PROPERTY, "Sierra Club");
+        sierraEntity.setProperty(GROUP_IMAGE_PROPERTY, "https://images.unsplash.com/photo-1584747167399-06a9ba" + 
+            "8302b0?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2704&q=80");
+        sierraEntity.setProperty(GROUP_DESCRIPTION_PROPERTY, "Group description would go here...");
+        datastore.put(sierraEntity);
+
+        Query groupsQuery = new Query(GROUP_KIND);
+        PreparedQuery preparedGroupsQuery = datastore.prepare(groupsQuery); 
+
+        ImmutableList.Builder<ImmutableMap<String, String>> builder = new ImmutableList.Builder<>();
+        for (Entity group : preparedGroupsQuery.asIterable()) {
+            ImmutableMap<String, String> groupMap = ImmutableMap.of(
+                GROUP_NAME_PROPERTY, (String) group.getProperty(GROUP_NAME_PROPERTY),
+                GROUP_IMAGE_PROPERTY, (String) group.getProperty(GROUP_IMAGE_PROPERTY),
+                GROUP_DESCRIPTION_PROPERTY, (String) group.getProperty(GROUP_DESCRIPTION_PROPERTY));
+            builder.add(groupMap);
```

Comment:
What exactly is the groupMap for? I see that you're using groupsList for the data for the template, is the groupMap part of that?

Reply:
yes, each group has a hashmap and then I put all those maps together in the list and pass that into the soy template